### PR TITLE
KEYCLOAK-14972 Add independent GroupProvider interface

### PR DESCRIPTION
**Pull request by mhajas**

### CODE REVIEW

**Review by martin-kanis — DISMISSED** [review]
@mhajas Thanks for the PR. It looks good to me. Approving.

File: server-spi-private/src/main/java/org/keycloak/models/GroupProviderFactory.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package org.keycloak.models;
```
Comment (sguilhen) [inline server-spi-private/src/main/java/org/keycloak/models/GroupProviderFactory.java:1]: Missing header

File: server-spi/src/main/java/org/keycloak/models/GroupProvider.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
+package org.keycloak.models;
```
Comment (sguilhen) [inline server-spi/src/main/java/org/keycloak/models/GroupProvider.java:1]: Missing header

File: server-spi/src/main/java/org/keycloak/models/GroupProvider.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    /**
+     * Returns groups with the given role in the given realm.
+     *
+     * @param realm Realm.
+     * @param role Role.
+     * @param firstResult First result to return. Ignored if negative.
+     * @param maxResults Maximum number of results to return. Ignored if negative.
+     * @return List of groups with the given role.
+     */
+    List<GroupModel> getGroupsByRole(RealmModel realm, RoleModel role, int firstResult, int maxResults);
```
Comment (sguilhen) [inline server-spi/src/main/java/org/keycloak/models/GroupProvider.java:71]: Shouldn't we have Stream variants for these methods?

**Review by sguilhen — CHANGES_REQUESTED** [review]
Overall it looks good @mhajas. Just spotted a few details.

File: server-spi/src/main/java/org/keycloak/models/GroupProvider.java (hunk 3 above)
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/models/GroupProvider.java:71]: This is going to be part of https://github.com/keycloak/keycloak/pull/7328

**Comment by mhajas** [comment]
@sguilhen Thank you for the review. I added the header to both files.

**Review by sguilhen — DISMISSED** [review]
LGTM

**Review by martin-kanis — DISMISSED** [review]

**Review by pedroigor — DISMISSED** [review]

**Comment by pedroigor** [comment]
@mhajas I can't merge due to Travis cross-dc failures.

**Comment by mhajas** [comment]
@pedroigor Seems like cross-dc tests passed.  Do you think the failures were caused by my PR or it is a known intermittent issue?

**Comment by pedroigor** [comment]
@mhajas They are stable. I had the same issues in several other PRs. In fact, this is the first time in months I see they passing so easily :)

**Review by pedroigor — APPROVED** [review]

**Pull request merged**